check-cadrum-occt-x86_64-pc-windows-gnu がホスト/コンテナ間 mingw libstdc++ ABI 不一致でリンク失敗

## 概要

`make check-cadrum-occt-x86_64-pc-windows-gnu` を Windows ホスト上で実行すると、最終リンク（ホスト側 `cargo run --example 01_primitives`）で undefined reference エラーになる。

コンテナ内での OCCT 静的ライブラリ生成（`cadrum-occt-x86_64-pc-windows-gnu` → `target/cadrum-occt-v800rc5-x86_64-pc-windows-gnu/win64/gcc/lib`）自体は成功している。失敗するのは抽出後にホスト側 mingw で最終リンクする `check-cadrum-occt-%` の `cargo run --example 01_primitives` ステップ。

## 再現コマンド

``` sh
make check-cadrum-occt-x86_64-pc-windows-gnu
```

## エラー抜粋

```
error: linking with `x86_64-w64-mingw32-gcc` failed: exit code: 1
  = note: C:/Users/0107411268/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/15.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe:
    libcadrum-...rlib(BinTools_ShapeSet.cxx.obj):
      undefined reference to `std::istream::seekg(std::fpos<int>)`
    libcadrum-...rlib(OSD_LocalFileSystem.cxx.obj):
      undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::seekpos(std::fpos<int>, std::_Ios_Openmode)`
  collect2.exe: error: ld returned 1 exit status
make: *** [makefile:22: check-cadrum-occt-x86_64-pc-windows-gnu] Error 101
```

## 原因の仮説

OCCT の `.cxx.obj` を固めた `.a` と最終リンク時の libstdc++ の ABI が違う。

- コンテナ側（`docker/Dockerfile_x86_64-pc-windows-gnu`）: Debian bookworm の `g++-mingw-w64-x86-64`（GCC 12 系, posix threads）で OCCT をコンパイル。
- ホスト側: `C:/Users/0107411268/mingw64` の **GCC 15.2.0**（MSYS2/WinLibs 等）で最終リンク。

`std::fpos<int>` / `std::istream::seekg` / `std::basic_filebuf::seekpos` のシンボル形は libstdc++ のバージョン／ABI 設定で変わり得るため、上記のような undefined reference は典型的な GCC バージョン跨ぎの ABI ミスマッチで発生する。

`check-cadrum-occt-%` の本来の意図は「コンテナで焼いた OCCT 静的ライブラリを、ホスト側で配布されている想定の mingw-gcc でリンクできるか検証する」ことのはずだが、両者の GCC バージョンを揃える仕組みがない。

## 対処方針案

1. コンテナ側の mingw GCC バージョンをホスト想定に揃える（bookworm の apt 既定ではなく、WinLibs 等を固定取得する Dockerfile に変更）。逆にホスト側を bookworm の 12 系に合わせる方針でもよい。
2. OCCT 側を `-static-libstdc++ -static-libgcc` で吐き、libstdc++ 依存を最終リンクから切り離す。ただし `link-cplusplus` / `cxx` 経由で引かれる C++ ランタイムとの整合確認が要る。
3. `check-cadrum-occt-%` 自体をコンテナ内で `cargo run` まで完結させる（ホスト側リンクを検証ターゲットから外す）。ただし `# cadrum-occt` tolerates the execution failure via `|| true` のコメント通り Wine 等が無いので実行は出来ず、"host でリンクできる" 検証の意図が損なわれる。

## 環境

- ホスト: Windows 11 Enterprise / mingw-w64 gcc 15.2.0
- コンテナ: `debian:bookworm-slim` + `g++-mingw-w64-x86-64`（posix alternative）
- cadrum: `main` @ 157e78b (v0.6.1)

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 target/
 out/
+logs/
 linkedin.md
 .venv/
 *.brep

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,7 +19,7 @@ pure = []
 # Build OCCT from upstream sources when the cache is empty, instead of
 # downloading a prebuilt tarball. Off by default — most users will use the
 # prebuilt binary path.
-source-build = ["dep:cmake", "dep:walkdir"]
+source-build = ["dep:cmake"]
 
 [dependencies]
 cxx = "1"
@@ -32,9 +32,9 @@ cxx-build = "1"
 minreq = { version = "2", features = ["https-rustls"] }
 libflate = "2"
 tar = "0.4"
+walkdir = "2"
 # source-build only: build OCCT from upstream sources
 cmake = { version = "0.1", optional = true }
-walkdir = { version = "2", optional = true }
 
 [package.metadata.docs.rs]
 targets = ["x86_64-pc-windows-msvc", "x86_64-unknown-linux-gnu"]

--- a/build.rs
+++ b/build.rs
@@ -119,6 +119,18 @@ fn link_occt_libraries(occt_include: &Path, occt_lib_dir: &Path) {
 		println!("cargo:rustc-link-lib=static={}", lib);
 	}
 
+	// Tarball 側が "cadrum" を含む名前の static library を同梱していれば拾う。
+	// 典型: mingw 向けに libcadrum_stdc++.a / libcadrum_gcc.a 等を同梱して、
+	// ホスト側 GCC のバージョン差による libstdc++ ABI ミスマッチを回避する (#89)。
+	// OCC_LIBS ループの後に置くので OCCT libs の未解決 symbol を後段で満たす順序になる。
+	for entry in walkdir::WalkDir::new(occt_lib_dir).min_depth(1).max_depth(1).into_iter().flatten() {
+		let Some(name) = entry.file_name().to_str() else { continue };
+		if name.contains("cadrum") {
+			let name=name.strip_prefix("lib").unwrap_or(name).strip_suffix(".a").or(name.strip_suffix(".lib")).unwrap_or(name);
+			println!("cargo:rustc-link-lib=static={}", name);
+		}
+	}
+
 	let target_env = env::var("CARGO_CFG_TARGET_ENV").unwrap_or_default();
 	let is_mingw_like = target_env == "gnu" || target_env == "gnullvm";
 	if is_mingw_like {

--- a/makefile
+++ b/makefile
@@ -10,9 +10,32 @@ deploy: generate # generate out/markdown from examples, then build out/html
 	./out/bin/mdbook build
 publish: deploy # publish to crates.io
 	cargo publish
+ifeq ($(CARGO_BUILD_TARGET),x86_64-pc-windows-gnu)
+# mingw gcc のインストール先から lib<name>.a の絶対 path を引く関数。
+# 使用例: $(call mingw_lib,libstdc++.a) → /usr/.../libstdc++.a
+# -print-file-name= はドライバの library search path から名前解決するだけで、C/C++ 区別はしないので
+# gcc 一本で stdc++/gcc/gcc_eh いずれも取れる (g++ でも同じ結果)。コンパイラは parse 時点で
+# 存在するので make 関数で OK (対 `$(shell find ...)` は parse 時評価なので cargo build 前に
+# 走り空を返す — OCCT lib dir の探索は shell substitution 側で行う、下の recipe を参照)。
+mingw_lib = $(shell x86_64-w64-mingw32-gcc -print-file-name=$(1))
+endif
+
 cadrum-occt: generate # build occt from source natively
 	cargo clean
 	cargo build --example 01_primitives --release --features source-build 2>&1 | tee out/log.txt # colorはdefaultの一部なのでfeature指定不要
+ifeq ($(CARGO_BUILD_TARGET),x86_64-pc-windows-gnu)
+	@# mingw コンテナで作った OCCT の lib dir にコンテナ側 gcc の runtime (libstdc++/libgcc/libgcc_eh) を
+	@# libcadrum_* リネームでコピーする。build.rs の scanner が libcadrum_* を自動で -l として拾い、
+	@# ホスト側 mingw との GCC バージョン差による ABI 不整合を回避する (#89 対策)。
+	@# libTKernel.a の位置から dirname で lib dir を得る。shell substitution `$$(...)` を使うのは
+	@# recipe 実行時に評価する必要があるため — make の `$(shell ...)` は recipe 開始時点で
+	@# 一括展開されるので cargo build 前に走って空を返してしまう。
+	@LIBDIR=$$(find target -maxdepth 6 -type f -name libTKernel.a -path '*cadrum-occt*' | head -n 1 | xargs -r dirname); \
+	[ -n "$$LIBDIR" ] || { echo "bundle: OCCT lib dir not found under target/" >&2; exit 1; }; \
+	cp -v "$(call mingw_lib,libstdc++.a)" "$$LIBDIR/libcadrum_stdc++.a"; \
+	cp -v "$(call mingw_lib,libgcc.a)"    "$$LIBDIR/libcadrum_gcc.a"; \
+	cp -v "$(call mingw_lib,libgcc_eh.a)" "$$LIBDIR/libcadrum_gcc_eh.a"
+endif
 	find target -maxdepth 1 -type d -name 'cadrum*' | xargs -IX sh -c 'tar -czf out/$$(basename X).tar.gz -C $$(dirname X) $$(basename X)'
 cadrum-occt-%: # build occt from source in cross ( = native build in container ) cadrum-occt-aarch64-unknown-linux-gnu cadrum-occt-x86_64-pc-windows-gnu cadrum-occt-x86_64-unknown-linux-gnu
 	docker build -f docker/Dockerfile_$(*) -t cadrum-occt-$(*) .

